src/codecs/hdr/decoder.rs

```diff
@@ -414,70 +428,56 @@ impl<R: BufRead> HdrDecoder<R> {
         let pixel_count = self.width as usize * self.height as usize;
         let mut ret = vec![Default::default(); pixel_count];
         for chunk in ret.chunks_mut(self.width as usize) {
-            read_scanline(&mut self.r, chunk)?;
+            read_scanline(&mut self.reader, chunk)?;
         }
         Ok(ret)
     }
 
-    /// Consumes decoder and returns a vector of transformed pixels
-    pub fn read_image_transform<T: Send, F: Send + Sync + Fn(Rgbe8Pixel) -> T>(
+    /// The callback gets a flattened index as the first argument.
+    pub fn read_image_transform<F: Send + Sync + FnMut(usize, Rgbe8Pixel)>(
         mut self,
-        f: F,
-        output_slice: &mut [T],
+        mut store_rgbe_pixel: F
     ) -> ImageResult<()> {
-        assert_eq!(
-            output_slice.len(),
-            self.width as usize * self.height as usize
-        );
-
-        // Don't read anything if image is empty
         if self.width == 0 || self.height == 0 {
             return Ok(());
         }
 
-        let chunks_iter = output_slice.chunks_mut(self.width as usize);
-        let mut pool = Pool::new(8); //
-
-        (pool.scoped(|scope| {
-            for chunk in chunks_iter {
-                let mut buf = vec![Default::default(); self.width as usize];
-                read_scanline(&mut self.r, &mut buf[..])?;
-                let f = &f;
-                scope.execute(move || {
```

Comment:
this code uses multi threading, but the compression is happening on one thread, while only copying the decompressed pixels is done on multiple threads. this seems like an accident. are there benchmarks that show us whether this multi threading speeds things up?

Reply:
without the threadpool:
```
load-Hdr/image1.hdr     time:   [356.12 us 361.68 us 367.43 us]
Found 10 outliers among 100 measurements (10.00%)
  7 (7.00%) high mild
  3 (3.00%) high severe

load-Hdr/rgbr4x4.hdr    time:   [1.7318 us 1.7542 us 1.7803 us]
Found 14 outliers among 100 measurements (14.00%)
  8 (8.00%) high mild
  6 (6.00%) high severe
```

Reply:
has been resolved in a different pr already